WordPress/gutenberg
# Refactor: useBlockTools hook

Closes https://github.com/WordPress/gutenberg/issues/57288 alternative to https://github.com/WordPress/gutenberg/pull/57291 and https://github.com/WordPress/gutenberg/pull/58947.

There are times where a block is selected but no tools are available. If we don't have any tools, we don't want to show a toolbar. This combines the logic for showing the block toolbar that was sprinkled throughout the codebase:

- block-toolbar/index.js checked if it should render based on if a toolbar was enabled for the block and if there were block controls in the slots
- block-tools/index.js determined if the popover should render based on cascading rules if the inserter popover or block select mode popover were showing (and other rules)
- use-can-block-toolbar-be-focused made an approximation of if the block toolbar would be showing or not, but fell short in some areas such as when a block toolbar should have been rendered, but there were no block tools available.



## What?

Unify logic for which block tools will be showing so there is a source of truth for what will be showing (block toolbar popover, fixed block toolbar, block inserter, block breadcrumb).

## Why?

Fixes https://github.com/WordPress/gutenberg/issues/57288.

## How?

- Adds a `useHasBlockToolbar` hook that has the logic for if the block-toolbar will render or not.
- Adds a `useBlockTools` hook that privately exports which block tools will be showing (block toolbar popover, fixed block toolbar, block inserter, block breadcrumb).
- Checks if any block tools are available before adding the `is-collapsed` class to the central toolbar area
- Removes selected block tools wrapper from DOM if no block tools available. I'm not sure if this was intentional to leave it in the DOM before, or if it was because there wasn't a reliable way of checking for it.

## Testing Instructions




Toolbar should show/hide in the same way as before.

### Testing Instructions for Keyboard

- Go to Site Editor
- Turn on Top Toolbar
- Edit Navigation Template Part
- Click navigation item
- Press Up Arrow until the block toolbar disappears
- Central command area should be present
- Central command area should not have `is-collapsed` class
- Alt + F10 should move focus to the document tools
